### ISSUE TITLE
MXFP8 GEMM rejects compact unswizzled scales that quantize_ accepts

### ISSUE THREAD

**Issue by wujingyue** [NVIDIA/TransformerEngine#3518]
For an MXFP8Tensor with `with_gemm_swizzled_scales=False`, `quantize_()` accepts compact scale buffers, but `general_gemm()` requires the caller to pad those buffers to GEMM tile dimensions before TE prepares them for GEMM.

For a 64×64 tensor, the logical scale shapes are `(64, 2)` rowwise and `(2, 64)` columnwise. Quantization succeeds with these actual allocations, preserves their shapes and storage, and produces the same logical data and scales as an independently quantized TE reference. GEMM then rejects them, expecting `(128, 4)` and `(4, 128)`, respectively. Using those padded allocations succeeds with the same data shape.

Expected behavior: GEMM should accept compact unswizzled scales and handle padding internally, either before swizzling or by reading the compact input directly into a padded, swizzled output. The GEMM-ready representation can retain its tile constraints without requiring the unswizzled storage representation to carry padding.

Reproducer (single Blackwell GPU; save as repro.py):

```python
import sys

import torch
import transformer_engine_torch as tex
from transformer_engine.pytorch.cpp_extensions import general_gemm
from transformer_engine.pytorch.tensor.mxfp8_tensor import MXFP8Quantizer, MXFP8Tensor

layout = sys.argv[1]  # TN or NN
padded = "--padded" in sys.argv[2:]
quantizer = MXFP8Quantizer(tex.DType.kFloat8E4M3)

def empty(shape):
    return torch.empty(shape, dtype=torch.uint8, device="cuda")

weight = MXFP8Tensor(
    shape=(64, 64),
    dtype=torch.bfloat16,
    device="cuda",
    rowwise_data=empty((64, 64)),
    columnwise_data=empty((64, 64)),
    rowwise_scale_inv=empty((128, 4) if padded else (64, 2)),
    columnwise_scale_inv=empty((4, 128) if padded else (2, 64)),
    fp8_dtype=tex.DType.kFloat8E4M3,
    quantizer=quantizer,
    with_gemm_swizzled_scales=False,
)
weight.quantize_(torch.randn((64, 64), dtype=torch.float32, device="cuda"))
torch.cuda.synchronize()
print("quantize_ succeeded", flush=True)

activation = quantizer(torch.randn((64, 64), dtype=torch.float32, device="cuda"))
general_gemm(weight, activation, out_dtype=torch.bfloat16, layout=layout)
torch.cuda.synchronize()
print("GEMM succeeded", flush=True)
```

Run each case in a separate process:

```bash
python repro.py TN           # Quantization succeeds; GEMM rejects rowwise scales.
python repro.py NN           # Quantization succeeds; GEMM rejects columnwise scales.
python repro.py TN --padded  # Both succeed.
python repro.py NN --padded  # Both succeed.
```

Observed errors (relevant excerpts):

```text
TN: transformer_engine/common/transformer_engine.cpp:103 in function CheckScaleTensorShape:
invalid scale_inv shape (expected (128,4), got (64,2))

NN: transformer_engine/common/transformer_engine.cpp:112 in function CheckScaleTensorShape:
invalid columnwise_scale_inv shape (expected (4,128), got (2,64))
```

Environment of the reproduced failure:

- GPU: NVIDIA GB200
- Transformer Engine: 2.18.0+27486e03
- PyTorch: 2.14.0a0+4fdf77b940.nv26.08
- CUDA reported by PyTorch: 13.4

Motivation: Megatron FSDP v2 stores and all-gathers compact, unswizzled MXFP8 planes, for memory. Requiring callers to pad scales before GEMM adds allocations and copies in the wrapper. TE already prepares unswizzled scales for GEMM, so it would be useful for that preparation to accept the compact representation directly.

**Comment by wujingyue** [comment]
I created #3519 as a draft for demonstration purposes. It would be great to also cover grouped GEMM and fuse padding into the swizzle operation—or, ideally, into the downstream GEMM.

I think an MXFP8Tensor should only require padded scales only when they’re swizzled. Since padding to 128×4 is only needed before GEMM, `with_gemm_swizzled_scales=True` is already a natural way to capture that requirement. Allowing unpadded scales when with_gemm_swizzled_scales=False would help avoid unnecessary padding and keep the API more intuitive.

**Comment by wujingyue** [comment]
Hey @vthumbe1503, 

I'm moving the discussion from https://github.com/NVIDIA/TransformerEngine/pull/3519#issuecomment-5686463645 to here. In general, I prefer to discuss use cases in issues, not PRs. 

> But in that case FSDP can handle padding those scale inv tensors after the shard.

You mean after the **unshard/allgather**? Yes, that's exactly how I've been **working around** this issue: https://github.com/wujingyue/Megatron-LM/blob/1d31d67138969fff686461686cdf11e4d1186e03/megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py#L208-L209. But it feels unnecessary, suboptimal and confusing. 

> How about we handle padding directly in the swizzling kernel, to avoid padding performance penalty?

Yes. That's exactly what I meant in https://github.com/NVIDIA/TransformerEngine/issues/3518#issuecomment-5686282202: "It would be great to also cover grouped GEMM and fuse padding into the swizzle operation—or, ideally, into the downstream GEMM."
